_posts/2020-05-18-mirror-kafka-eventhub.md

```diff
@@ -0,0 +1,225 @@
+    groupId: my-source-group-id
+  producer:
+    bootstrapServers: <eventhubs-namespace>.servicebus.windows.net:9093
+    authentication:
+      type: plain
+      username: $ConnectionString
+      passwordSecret:
+        secretName: eventhubssecret
+        password: eventhubspassword
+    tls:
+      trustedCertificates: []
+  whitelist: ".*"
+```
+
+The `consumer` part just connects to the local Apache Kafka cluster running on the same Kubernetes instance.
+
+The `producer` part connects to the Event Hub namespace, through TLS and PLAIN authentication using `$ConnectionString` as username and the connection string as password provided in the already created `eventhubssecret`.
+The `tls` section is used because Event Hub connection [needs SSL](https://docs.microsoft.com/en-gb/azure/event-hubs/event-hubs-for-kafka-ecosystem-overview?WT.mc_id=devto-blog-abhishgu#security-and-authentication) with `SASL_SSL` as security protocol
+
+Create the `KafkaMirrorMaker` on the Kubernetes cluster saving the above resource into a `kafka-mirror-maker.yaml` file.
+The Strimzi cluster operator takes care of it deploying Kafka Mirror Maker using the above configuration.
+
+```shell
+kubectl apply -f kafka-mirror-maker.yaml -n kafka
+```
+
+### Produce, mirror and trigger!
+
+The application consuming the mirrored messages from Event Hub is a simple Azure Functions application which just gets and logs the message that is enough for the purpose of this demo.
+
+```java
+/**
+ * Azure Functions with Event Hub trigger.
+ */
+public class StrimziEventHubTrigger {
+    /**
+     * This function will be invoked when an event is received from Event Hub.
+     */
+    @FunctionName("StrimziEventHubTrigger")
+    public void run(
+        @EventHubTrigger(name = "message", eventHubName = "<eventhubs-namespace>", connection = "EH_CONNECTION_STRING", consumerGroup = "$Default", cardinality = Cardinality.MANY) List<String> message,
+        final ExecutionContext context
+    ) {
+        context.getLogger().info("Java Event Hub trigger function executed.");
+        context.getLogger().info("Length:" + message.size());
+        message.forEach(singleMessage -> context.getLogger().info(singleMessage));
+    }
+}
+```
+
+Even in this case the `<eventhubs-namespace>` has to be replaced with the Event Hub namespace to connect to and the `EH_CONNECTION_STRING` environment variable has to be defined in the JSON application settings file containing the connection string as already used for the Kafka Mirror Maker configuration.
+
+For more information about Azure Functions connecting to Event Hub this [link](https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-event-hubs-trigger) provides really useful information.
+The application can run even locally or be published on the Azure cloud.
+
+To try the entire pipeline, we just miss sending messages and see how they are mirrored to Event Hub triggering the above function to be executed.
```

Comment:
just miss sending messages?

"...causing the above function..." would be better than triggering. The "triggering" and the "executed" is a bit tautologous.

Reply:
I meant that the part we haven't covered yet is sending messages to see how they are mirrored and Azure Functions application is triggered.

Reply:
> just miss sending messages?

@tombentley what did you mean by that?

Reply:
I think @tombentley means that it would read better as "To try the entire pipeline, the only thing left to do is to send some messages. Then we can see that they are mirrored to Event Hub, causing the above function to be executed".

Reply:
@ppatierno sorry I wasn't clear. I meant that I didn't really understand what you were meaning. Simon's suggestion SGTM.